lib/ziti_tunnel_ctrl.c

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2019 Netfoundry, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <ziti/ziti_tunnel_cbs.h>
+#include <ziti/ziti_log.h>
+
+IMPL_ENUM(TunnelCommand, TUNNEL_COMMANDS)
+
+static struct cmd_ctx_s {
+    ziti_tunnel_ctrl ctrl;
+    tunneler_context tunnel_ctx;
+    command_cb cb;
+    uv_loop_t *loop;
+} CMD_CTX;
+
+struct ziti_instance_s {
+    ziti_options opts;
+    command_cb load_cb;
+    void *load_ctx;
+
+    ziti_context ztx;
+    LIST_ENTRY(ziti_instance_s) _next;
+};
+
+// temporary list to pass info between parse and run
+static LIST_HEAD(instance_list, ziti_instance_s) instance_init_list;
+
+// map<path -> ziti_instance>
+static model_map instances;
+
+static void on_ziti_event(ziti_context ztx, const ziti_event_t *event);
+
+static const char * cfg_types[] = { "ziti-tunneler-client.v1", "intercept.v1", "ziti-tunneler-server.v1", "host.v1", NULL };
+
+static long refresh_interval = 10;
+
+static int process_cmd(const tunnel_comand *cmd, void (*cb)(const tunnel_result *, void *ctx), void *ctx);
+static int load_identity(const char *path, command_cb cb, void *ctx);
+static struct ziti_instance_s *new_ziti_instance(const char *path);
+static void load_ziti_async(uv_async_t *ar);
+
+const ziti_tunnel_ctrl* ziti_tunnel_init_cmd(uv_loop_t *loop, tunneler_context tunnel_ctx, command_cb cb) {
+    CMD_CTX.loop = loop;
+    CMD_CTX.tunnel_ctx = tunnel_ctx;
+    CMD_CTX.cb = cb;
+    CMD_CTX.ctrl.process = process_cmd;
+    CMD_CTX.ctrl.load_identity = load_identity;
+    return &CMD_CTX.ctrl;
+}
+
+static int process_cmd(const tunnel_comand *cmd, command_cb cb, void *ctx) {
+    tunnel_result result = {
+                    .success = false,
+                    .error = NULL,
+                    .data = NULL,
+    };
+    ZITI_LOG(INFO, "processing command[%s] with data[%s]", TunnelCommands.name(cmd->command), cmd->data);
+    switch (cmd->command) {
```

Comment:
should this switch have a case for `TunnelCommand_ZitiDump`?

Reply:
sure, once we implement it